src/dashboardGenerator.js

```diff
@@ -0,0 +1,359 @@
+  const maxMetricsPerWidget = 100;
+  const maxMetricsPerServiceInAWidget = 2;
+  const maxServices = Math.floor(maxMetricsPerWidget / maxMetricsPerServiceInAWidget);
+
+  // For each service there will be three metrics (MTTR, MTBF, MTTF)
+  const maxSupportedMetrics = maxServices * 3;
+
+  if (state.metrics.length > maxSupportedMetrics) {
+    console.warn(`Maximum of ${maxSupportedMetrics} metrics are allowed in a single dashboard. Some metrics will not be reported.`);
+    state.metrics = state.metrics.sort((a, b) => (a.serviceName > b.serviceName) ? 1 : -1);
+    state.metrics = state.metrics.slice(0, maxSupportedMetrics);
+  }
+};
+
+function trendWidgets(metrics, y, state) {
+
+  return state.widgetMappings.map(mapping => {
+    const region = state.region;
+    const filterredGroup = metrics.filter(x => x.metricName == mapping.label);
+    let resultMetrics = filterredGroup.map((mappinggroup, index) => {
+      return [
+        [
+          {
+            "expression": `FILL(m${index}/${mapping.unitConversion.unit},AVG(m${index}/${mapping.unitConversion.unit}))`,
+            "id": `e${index}`,
+            "region": "ap-southeast-2",
+            "label": mappinggroup.serviceName
+          }
+        ],
+        [
+          "Operations",
+          mapping.label,
+          "service",
+          mappinggroup.serviceName,
+          {
+            "stat": "Average",
+            "id": `m${index}`,
+            "visible": false,
+            "period": SEVEN_DAYS
+          }
+        ]
+      ]
+    });
+    //flatten the arrays
+    resultMetrics = [].concat.apply([], resultMetrics);
+
+    return {
+      "type": "metric",
+      "x": mapping.x,
+      "y": mapping.y,
+      "width": WIDGET_WIDTH,
+      "height": WIDGET_HEIGHT,
+      "properties": {
+        "metrics": resultMetrics,
+        "view": "timeSeries",
+        "region": region,
+        "title": mapping.label + ' Trend',
+        "period": ONE_DAYS,
+        "stacked": false,
+        "yAxis": {
+          "left": {
+            "min": 0,
+            "label": mapping.unitConversion.label,
+            "showUnits": false
+          },
+          "right": {
+            "showUnits": true
+          }
+        },
+        "annotations": mapping.properties.annotations,
+        "stat": "Sum"
+      }
+    };
+  });
+
+}
+
+function averageWidgets(metrics, y, state) {
+
+  return state.widgetMappings.map(mapping => {
+
+    const filterredGroup = metrics.filter(x => x.metricName == mapping.label);
+    let resultMetrics = filterredGroup.map((mappingGroup, index) => {
+      return [
+        "Operations",
+        mappingGroup.metricName,
+        "service",
+        mappingGroup.serviceName,
+        {
+          "id": `m${index}`,
+          "label": mappingGroup.serviceName
+        }
+      ]
+    });
+    return {
+      "type": "metric",
+      "x": mapping.x,
+      "y": 10,
```

Comment:
See comment above:
> The relationship between location and the size of the widget is in 2 places:

Actually it's 3 places - including here.

Can we manage the relationship/calculation in one place?

Reply:
How much effort would that be?

Reply:
Done